cns/ipampoolmonitor/ipampoolmonitor.go

```diff
@@ -2,130 +2,197 @@
+	// TODO: Better handling here for negatives
+	pm.cachedSpec.RequestedIPCount -= pm.scalarUnits.BatchSize
 
 	// mark n number of IP's as pending
-	pendingIPAddresses, err := pm.cns.MarkIPsAsPending(decreaseIPCount)
+	pendingIPAddresses, err := pm.cns.MarkIPsAsPending(int(pm.scalarUnits.BatchSize))
 	if err != nil {
 		return err
 	}
 
 	// convert the pending IP addresses to a spec
-	spec, err := CNSToCRDSpec(pendingIPAddresses, decreaseIPCount)
+	pm.cachedSpec, err = CNSToCRDSpec(pendingIPAddresses, pm.cachedSpec.RequestedIPCount)
 	if err != nil {
 		return err
 	}
+	pm.pendingRelease = true
+	logger.Printf("[ipam-pool-monitor] Decreasing pool size, Current Pool Size: %v, Requested IP Count: %v, Pods with IP's: %v", len(pm.cns.GetPodIPConfigState()), pm.cachedSpec.RequestedIPCount, len(pm.cns.GetAllocatedIPConfigs()))
+	return pm.rc.UpdateCRDSpec(context.Background(), pm.cachedSpec)
+}
+
+// if cns pending ip release map is empty, request controller has already reconciled the CNS state,
+// so we can remove it from our cache and remove the IP's from the CRD
+func (pm *CNSIPAMPoolMonitor) cleanPendingRelease() error {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+
+	var err error
+	pm.cachedSpec, err = CNSToCRDSpec(nil, pm.cachedSpec.RequestedIPCount)
+	if err != nil {
+		logger.Printf("[ipam-pool-monitor] Failed to translate ")
+	}
 
-	return pm.rc.UpdateCRDSpec(context.Background(), spec)
+	pm.pendingRelease = false
+	return pm.rc.UpdateCRDSpec(context.Background(), pm.cachedSpec)
 }
 
 // CNSToCRDSpec translates CNS's map of Ips to be released and requested ip count into a CRD Spec
-func CNSToCRDSpec(toBeDeletedSecondaryIPConfigs map[string]cns.SecondaryIPConfig, ipCount int) (nnc.NodeNetworkConfigSpec, error) {
+func CNSToCRDSpec(toBeDeletedSecondaryIPConfigs map[string]cns.IPConfigurationStatus, ipCount int64) (nnc.NodeNetworkConfigSpec, error) {
 	var (
 		spec nnc.NodeNetworkConfigSpec
 		uuid string
 	)
 
-	spec.RequestedIPCount = int64(ipCount)
+	spec.RequestedIPCount = ipCount
 
-	for uuid = range toBeDeletedSecondaryIPConfigs {
-		spec.IPsNotInUse = append(spec.IPsNotInUse, uuid)
+	if toBeDeletedSecondaryIPConfigs == nil {
+		spec.IPsNotInUse = make([]string, 0)
+	} else {
+		for uuid = range toBeDeletedSecondaryIPConfigs {
+			spec.IPsNotInUse = append(spec.IPsNotInUse, uuid)
+		}
 	}
 
 	return spec, nil
 }
+
+// UpdatePoolLimitsTransacted called by request controller on reconcile to set the batch size limits
+func (pm *CNSIPAMPoolMonitor) UpdatePoolMonitor(scalarUnits cns.ScalarUnits) error {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	pm.scalarUnits = scalarUnits
+
+	pm.MinimumFreeIps = int64(float64(pm.scalarUnits.BatchSize) * (float64(pm.scalarUnits.RequestThresholdPercent) / 100))
+	pm.MaximumFreeIps = int64(float64(pm.scalarUnits.BatchSize) * (float64(pm.scalarUnits.ReleaseThresholdPercent) / 100))
+
+	if pm.cns == nil {
+		return fmt.Errorf("Error Updating Pool Limits, reference to CNS is nil")
+	}
+
```

Comment:
Initial RequestedIpCount should be same as what is specified in the spec. Say you updated the CRP spec RequestedCount to 20 and then CNS died. On reconcile we want to make sure we cache the CRD spec. (Both RequestedCount + IPs which are marked as pending release)